velox/exec/Spill.h

```diff
@@ -0,0 +1,362 @@
+  // the file system path prefix. 'bits' is the hash bit field for
+  // partitioning data between files. This also gives the maximum
+  // number of partitions. 'numSortingKeys' is the number of leading columns
+  // on which the data is sorted, 0 if only hash partitioning is used.
+  // 'targetFileSize' is the target size of a single
+  // file. 'targetBatchSize is the target number of rows in a single
+  // RowVector written to a spill file.  'pool' and 'mappedMemory' own
+  // the memory for state and results.
+  SpillState(
+      const std::string& path,
+      HashBitRange bits,
+      int32_t numSortingKeys,
+      uint64_t targetFileSize,
+      uint64_t targetBatchSize,
+      memory::MemoryPool& pool,
+      memory::MappedMemory& mappedMemory)
+      : path_(path),
+        hashBits_(bits),
+        fieldMask_(((1UL << (hashBits_.end - hashBits_.begin))) - 1),
+        numSortingKeys_(numSortingKeys),
+        targetFileSize_(targetFileSize),
+        targetBatchSize_(targetBatchSize),
+        pool_(pool),
+        mappedMemory_(mappedMemory) {}
+
+  // Returns which spill partition 'hash' falls into. Returns -1 if the
+  // partition of 'hash' has not been started.
+  int32_t partition(uint64_t hash) const {
+    int32_t field = (hash >> hashBits_.begin) & fieldMask_;
+    return field < numPartitions_ ? field : -1;
+  }
+
+  int32_t numPartitions() const {
+    return numPartitions_;
+  }
+
+  // Sets how many of the spill partitions are in use.
+  void setNumPartitions(int32_t numPartitions);
+
+  // Returns how many ways spilled data can be partitioned.
+  int32_t maxPartitions() const {
+    return 1 << (hashBits_.end - hashBits_.begin);
+  }
+
+  uint64_t targetFileSize() const {
+    return targetFileSize_;
+  }
+
+  uint64_t targetBatchSize() const {
+    return targetBatchSize_;
+  }
+
+  memory::MemoryPool& pool() const {
+    return pool_;
+  }
+
+  // Appends data to 'partition'. The rows  given by 'indices'  must be sorted
+  // for a sorted spill and must hash to 'partition'.
+  void appendToPartition(int32_t partition, const RowVectorPtr& rows);
+
+  // Finishes a sorted run for 'partition'. If write is called for 'partition'
+  // again, the data does not have to be sorted relative to the data
+  // written so far.
+  void finishWrite(int32_t partition) {
+    files_[partition]->finishFile();
+  }
+
+  // Starts reading values for 'partition'. If 'extra' is non-null, it can be
+  // a stream of rows from a RowContainer so as to merge unspilled
+  // data with spilled data.
+  std::unique_ptr<TreeOfLosers<SpillStream>> startMerge(
+      int32_t partition,
+      std::unique_ptr<SpillStream>&& extra);
+
+ private:
+  const RowTypePtr type_;
+  const std::string path_;
+  const HashBitRange hashBits_;
+  const uint64_t fieldMask_;
+  const int32_t numSortingKeys_;
+  // Number of spilled ranges. This is 2** the number of bits in the range of
+  // 'hashBits_'
+  int32_t numPartitions_ = 0;
```

Comment:
numPartitions_ appears to mutable, which contracts the comment that says "This is 2** the number of bits in the range of 'hashBits_'". Why is numPartitions_ mutable? Why is there setNumPartitions() method?